modules/ipp/scanner.go

```diff
@@ -0,0 +1,206 @@
+}
+
+// Module implements the zgrab2.Module interface.
+type Module struct {
+	// TODO: Add any module-global state if necessary
+}
+
+// Scanner implements the zgrab2.Scanner interface.
+type Scanner struct {
+	config *Flags
+	// TODO: Add scan state if any is necessary
+}
+
+// RegisterModule registers the zgrab2 module.
+func RegisterModule() {
+	var module Module
+	_, err := zgrab2.AddCommand("ipp", "ipp", "Probe for ipp", 631, &module)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+// NewFlags returns a default Flags object.
+func (module *Module) NewFlags() interface{} {
+	return new(Flags)
+}
+
+// NewScanner returns a new Scanner instance.
+func (module *Module) NewScanner() zgrab2.Scanner {
+	return new(Scanner)
+}
+
+// Validate checks that the flags are valid.
+// On success, returns nil.
+// On failure, returns an error instance describing the error.
+func (flags *Flags) Validate(args []string) error {
+	return nil
+}
+
+// Help returns the module's help string.
+func (flags *Flags) Help() string {
+	//TODO: Write a help string
+	return ""
+}
+
+// Init initializes the Scanner.
+func (scanner *Scanner) Init(flags zgrab2.ScanFlags) error {
+	f, _ := flags.(*Flags)
+	scanner.config = f
+	//TODO: Take action in response to flags which were set
+	return nil
+}
+
+// InitPerSender initializes the scanner for a given sender.
+func (scanner *Scanner) InitPerSender(senderID int) error {
+	return nil
+}
+
+// GetName returns the Scanner name defined in the Flags.
+func (scanner *Scanner) GetName() string {
+	return scanner.config.Name
+}
+
+// Protocol returns the protocol identifier of the scan.
+func (scanner *Scanner) Protocol() string {
+	return "ipp"
+}
+
+// GetPort returns the port being scanned.
+func (scanner *Scanner) GetPort() uint {
+	return scanner.config.Port
+}
+
+//FIXME: Maybe switch to ipp/ipps schemes, at least optionally
+func getIPPURL(https bool, host string, port uint16, endpoint string) string {
+	var proto string
+	if https {
+		proto = "https"
+	} else {
+		proto = "http"
+	}
+	return proto + "://" + host + ":" + strconv.FormatUint(uint64(port), 10) + endpoint
+}
+
+func ippInContentType(resp http.Response) bool {
+	for _, t := range resp.Header["Content-Type"] {
+		if strings.Contains(t, CONTENT_TYPE) {
+			return true
+		}
+	}
+	return false
+}
+
+//TODO: Doesn't support TLS at all right now
+func (scanner *Scanner) grab(target zgrab2.ScanTarget) (int8, int8, *zgrab2.ScanError) {
+	//FIXME: This is not where this hostname assignment logic should live
+	host := target.Domain
+	if host == "" {
+		host = target.IP.String()
+	}
+	//FIXME: ?Should just use endpoint "/", since we get the same response as "/ipp" on CUPS??
+	uri := getIPPURL(scanner.config.IPPSecure, host, uint16(scanner.config.BaseFlags.Port), "/ipp")
+	b := getPrinterAttributesRequest(uri)
+	resp, err := http.Post(uri, CONTENT_TYPE, &b)
+	if err != nil {
+		//TODO: Create a descriptive error
```

Comment:
You can use `zgrab2.TryGetScanStatus(err)` to auto-detect certain statuses, like timeouts / connection refused. This would be a good place to use that.

 - [ ] Detect the status